### PR TITLE
 protoc-gen-swagger: Add ability to specify custom response objects 

### LINKED ISSUE #304: Response codes and descriptions in Swagger docs
It would be nice if we can specify rest response codes/descriptions via some grpc gateway annotations or some other means.

### PR DESCRIPTION
**Pull request by johanbrandhorst**
Extends the openapiv2 protofile to allow control over response objects.

Fixes #304

### CODE REVIEW

**Comment by johanbrandhorst** [comment]
@ivucica @tmc @achew22

**Comment by johanbrandhorst** [comment]
Not sure I understand the test failures, do I need to regenerate the file?

**Comment by johanbrandhorst** [comment]
Ok I hope to have fixed it again

**Comment by achew22** [comment]
I like this PR, I need to look at it in a bigger screen then my phone but it looks like a great additional piece of documentation to include

File: examples/proto/examplepb/a_bit_of_everything.proto [hunk 1]
```diff
@@ -11,89 +11,106 @@ import "google/protobuf/timestamp.proto";
...
-	  }
-	  scope: {
-	    key: "write";
-	    value: "Grants write access";
-	  }
-	  scope: {
-	    key: "admin";
-	    value: "Grants read and write access to administrative information";
-	  }
+	info: {
```
Comment (achew22) [inline examples/proto/examplepb/a_bit_of_everything.proto]: Out of curiosity, why did you change everything to be tabs? Can you run this through a modern version of clang-format?

Comment (johanbrandhorst) [reply examples/proto/examplepb/a_bit_of_everything.proto]: Everything except for the examples added by @ivucica was already tabs, so I just made it consistent as it was driving me mad. I could format this (and the other protofiles) with `prototool format` [1] if you like, but I havne't touched the others.

[1]: https://github.com/uber/prototool#prototool-format

**Comment by johanbrandhorst** [comment]
Gonna see if I can add automatic references for namespaced messages as well, so please don't merge just yet.

**Comment by johanbrandhorst** [comment]
Actually, go ahead and merge if it's good, I'll make another PR for the additional functionality.

**Comment by johanbrandhorst** [comment]
Wasn't too hard it runs out :)

File: protoc-gen-swagger/genswagger/template.go [hunk 2]
```diff
@@ -1187,3 +1230,70 @@ func extractSwaggerOptionFromFileDescriptor(file *pbdescriptor.FileDescriptorPro
...
+		// TODO(johanbrandhorst): Add more fields?
+	}
+	if s.GetJsonSchema().GetRef() != "" {
+		swaggerName := fullyQualifiedNameToSwaggerName(s.GetJsonSchema().GetRef(), reg)
+		if swaggerName != "" {
+			ret.schemaCore.Ref = "#/definitions/" + swaggerName
+		} else {
+			ret.schemaCore.Ref += s.GetJsonSchema().GetRef()
+		}
+	} else {
```
Comment (johanbrandhorst) [inline protoc-gen-swagger/genswagger/template.go]: One thing this doesn't do is include external messages that aren't already inlined. I think we could potentially do some proto registry lookups and add those messages if necessary, but it'd be quite complex so I'm not sure we want to?

File: examples/proto/examplepb/a_bit_of_everything.proto (hunk 1 above)
Comment (ivucica) [reply examples/proto/examplepb/a_bit_of_everything.proto]: (Sorry about this.)

File: examples/proto/examplepb/a_bit_of_everything.proto [hunk 3]
```diff
@@ -155,7 +183,7 @@ message ABitOfEverything {
 
 	google.protobuf.Timestamp timestamp_value = 27;
 
-        // repeated enum value. it is comma-separated in query
+		// repeated enum value. it is comma-separated in query
```
Comment (ivucica) [inline examples/proto/examplepb/a_bit_of_everything.proto]: Remove one tab maybe?

File: protoc-gen-swagger/genswagger/template.go [hunk 4]
```diff
@@ -1187,3 +1230,70 @@ func extractSwaggerOptionFromFileDescriptor(file *pbdescriptor.FileDescriptorPro
 	}
 	return opts, nil
 }
+
+func swaggerSchemaFromProtoSchema(s *swagger_options.Schema, reg *descriptor.Registry) swaggerSchemaObject {
```
Comment (ivucica) [inline protoc-gen-swagger/genswagger/template.go]: Should https://github.com/johanbrandhorst/grpc-gateway/blob/97aefd3e41117e62982ef489dfb516a228a39ea3/protoc-gen-swagger/genswagger/template.go#L231-L245 and other places extracting the schema object be combined with this?

File: protoc-gen-swagger/genswagger/template.go [hunk 5]
```diff
@@ -1187,3 +1230,70 @@ func extractSwaggerOptionFromFileDescriptor(file *pbdescriptor.FileDescriptorPro
...
+	case swagger_options.JSONSchema_STRING:
+		// NOTE: in swagger specifition, format should be empty on string type
+		return "string", ""
+	default:
+		// Maybe panic?
+		return "", ""
+	}
+}
+
+func protoExternalDocumentationToSwaggerExternalDocumentation(in *swagger_options.ExternalDocumentation) *swaggerExternalDocumentationObject {
```
Comment (ivucica) [inline protoc-gen-swagger/genswagger/template.go]: Should https://github.com/johanbrandhorst/grpc-gateway/blob/97aefd3e41117e62982ef489dfb516a228a39ea3/protoc-gen-swagger/genswagger/template.go#L233-L241 and other places with `ExternalDocs` be combined with this?

File: protoc-gen-swagger/genswagger/template_test.go [hunk 6]
```diff
@@ -4,12 +4,13 @@ import (
 	"reflect"
 	"testing"
 
+	"fmt"
+
```
Comment (ivucica) [inline protoc-gen-swagger/genswagger/template_test.go]: Actually if this is being moved, maybe move it together with `"testing"` and `"reflect"` and others? Seems strange to have it hang on its own

File: protoc-gen-swagger/options/openapiv2.proto [hunk 7]
```diff
@@ -132,8 +147,13 @@ message JSONSchema {
   reserved 1;
   // field 2 is reserved for '$schema', omitted from OpenAPI v2.
   reserved 2;
-  // field 3 is reserved for '$ref', although it is unclear how it would be used.
-  reserved 3;
+  // Ref is used to define an external reference to include in the message.
+  // This should be a fully qualified message reference, and that type must be imported
```
Comment (ivucica) [inline protoc-gen-swagger/options/openapiv2.proto]: > qualified message reference

I'd be slightly confused about this, maybe `fully qualified protobuf message reference`? Not sure.

Comment (ivucica) [inline protoc-gen-swagger/options/openapiv2.proto]: > This should be

This *could* be, I suppose?

File: examples/proto/examplepb/a_bit_of_everything.proto (hunk 3 above)
Comment (johanbrandhorst) [reply examples/proto/examplepb/a_bit_of_everything.proto]: Woops! Removed.

File: protoc-gen-swagger/genswagger/template.go (hunk 4 above)
Comment (johanbrandhorst) [reply protoc-gen-swagger/genswagger/template.go]: Great idea, I'll see if I can make that work.

File: protoc-gen-swagger/genswagger/template_test.go (hunk 6 above)
Comment (johanbrandhorst) [reply protoc-gen-swagger/genswagger/template_test.go]: Didn't even realise I had changed this, thanks linters!

File: protoc-gen-swagger/options/openapiv2.proto (hunk 7 above)
Comment (johanbrandhorst) [reply protoc-gen-swagger/options/openapiv2.proto]: I've adjusted it accordingly. I hope together with the example it is obvious enough.

File: protoc-gen-swagger/genswagger/template.go (hunk 5 above)
Comment (johanbrandhorst) [reply protoc-gen-swagger/genswagger/template.go]: I think I squashed these.

File: protoc-gen-swagger/genswagger/template.go [hunk 8]
```diff
@@ -229,19 +229,19 @@ func renderMessagesAsDefinition(messages messageMap, d swaggerDefinitionsObject,
...
+			schema.ExternalDocs = protoSchema.ExternalDocs
+			if protoSchema.schemaCore.Type != "" || protoSchema.schemaCore.Ref != "" {
+				schema.schemaCore = protoSchema.schemaCore
+			}
+			if protoSchema.Title != "" {
+				schema.Title = protoSchema.Title
+			}
+			if protoSchema.Description != "" {
+				schema.Description = protoSchema.Description
+			}
```
Comment (johanbrandhorst) [inline protoc-gen-swagger/genswagger/template.go]: I added these because I need them :P

**Comment by johanbrandhorst** [comment]
@ivucica are you happy with the changes? @achew22 could you rerun the failure please, seems like a github TLS timeout error?

**Comment by achew22** [comment]
Retriggered

**Comment by achew22** [comment]
I think the error comes from an update to the protobuf generator. Could you update your protoc and go plugin and regenerate with `make examples` then upload again?

I expect it should add the following

```patch
index cb6fb3a..d416004 100644
--- a/examples/clients/abe/examplepb_a_bit_of_everything.go
+++ b/examples/clients/abe/examplepb_a_bit_of_everything.go
@@ -14,6 +14,7 @@ import (
 	"time"
 )
 
+// Intentionaly complicated message type to cover much features of Protobuf.
 type ExamplepbABitOfEverything struct {
 
 	SingleNested ABitOfEverythingNested `json:"single_nested,omitempty"`
 ```

**Comment by ivucica** [comment]
No objections on my part as long as the tests are fixed.

**Comment by johanbrandhorst** [comment]
Cool, I think I should have fixed the failing test, lets get this is :D.

**Comment by achew22** [comment]
Looks like there are a few remaining issues. Can you give it another whack?

**Comment by johanbrandhorst** [comment]
@achew22 I can't seem to get swagger-codegen installed - I hate to be a bother but any chance you could just clone and regenerate the examples? I'll take another tomorrow at work but haven't got time now.

**Comment by ivucica** [comment]
Last time I was doing it, I just did what was done for Travis. :-)

> --
Sent from Gmail Mobile

**Comment by johanbrandhorst** [comment]
@ivucica thanks I will take a look through the logs. I tried using their docker image and it didn't work.

**Comment by johanbrandhorst** [comment]
Alright I should've gotten them this time. Sure is fiddly :thinking:.

**Comment by johanbrandhorst** [comment]
OK so I had some time to look into recursive resolution of manually referenced types - I got it working but I can't help but feel like it's a bit hacky. I'd be happy with eny tips on improving it, but basically it required me to keep a separate `set` of referenced definitions and render them after everything else, recursively.

**Comment by johanbrandhorst** [comment]
Bump - any thoughts on the additions?

File: examples/clients/abe/examplepb_a_bit_of_everything.go [hunk 9]
```diff
@@ -14,6 +14,7 @@ import (
 	"time"
 )
 
+// Intentionaly complicated message type to cover much features of Protobuf.
```
Comment (ivucica) [inline examples/clients/abe/examplepb_a_bit_of_everything.go]: Nit: s/much/many/

File: examples/proto/examplepb/a_bit_of_everything.proto [hunk 10]
```diff
@@ -94,13 +94,45 @@ option (grpc.gateway.protoc_gen_swagger.options.openapiv2_swagger) = {
...
 };
 
 
 // Intentionaly complicated message type to cover much features of Protobuf.
 // NEXT ID: 30
 message ABitOfEverything {
 	option (grpc.gateway.protoc_gen_swagger.options.openapiv2_schema) = {
+		json_schema: {
+			title: "A bit of everything"
+			description: "Intentionaly complicated message type to cover much features of Protobuf."
```
Comment (ivucica) [inline examples/proto/examplepb/a_bit_of_everything.proto]: Nit: s/much/many/

We can skip this if it'd take you a lot of time to regenerate examples.

File: protoc-gen-swagger/genswagger/template.go [hunk 11]
```diff
@@ -512,7 +515,7 @@ func templateToSwaggerPath(path string) string {
 	return strings.Join(parts, "/")
 }
 
-func renderServices(services []*descriptor.Service, paths swaggerPathsObject, reg *descriptor.Registry, refs refMap) error {
+func renderServices(services []*descriptor.Service, paths swaggerPathsObject, reg *descriptor.Registry, requestResponseRefs refMap, customRefs refMap) error {
```
Comment (ivucica) [inline protoc-gen-swagger/genswagger/template.go]: `requestResponseRefs, customRefs refMap` is also valid

File: protoc-gen-swagger/genswagger/template.go [hunk 12]
```diff
@@ -726,17 +727,18 @@ func applyTemplate(p param) (*swaggerObject, error) {
...
+	// Also adds custom user specified references to second map.
+	requestResponseRefs, customRefs := refMap{}, refMap{}
+	if err := renderServices(p.Services, s.Paths, p.reg, requestResponseRefs, customRefs); err != nil {
 		panic(err)
 	}
 
-	// Find all the service's messages and enumerations that are defined (recursively) and then
-	// write their request and response types out as definition objects.
+	// Find all the service's messages and enumerations that are defined (recursively)
+	// and write request, response and referenced types out as definition objects.
```
Comment (ivucica) [inline protoc-gen-swagger/genswagger/template.go]: Maybe: s/referenced types/other custom (but referenced) types/ ?

File: examples/clients/abe/examplepb_a_bit_of_everything.go (hunk 9 above)
Comment (johanbrandhorst) [reply examples/clients/abe/examplepb_a_bit_of_everything.go]: Yeh I intentionally didn't change this existing typo but may as well :D

**Comment by johanbrandhorst** [comment]
@tmc @achew22 I think the test failure was a flaky test, could one of you rerun it please?

**Comment by achew22** [comment]
Rerun. I'll try to check on it in 20 but I'll probably forget. If I don't do something by tomorrow morning please reply here, it means I forgot (sorry in advance)

**Comment by johanbrandhorst** [comment]
@achew22 thanks for retriggering it, seems like it passed ;). I appreciate the time you put into maintaining this project, so no worries about forgetting!

**Comment by johanbrandhorst** [comment]
@ivucica anymore comments?

**Comment by johanbrandhorst** [comment]
Friendly Monday bump @achew22 @ivucica

**Comment by achew22** [comment]
LGTM. Let's give @ivucica 24 hours from your ping before merging. I will rebase tomorrow around this time. If I don't please reping and I'll merge when I get to a computer.

PS: Sorry this is takling forever, it's a great feature and I am REALLY looking forward to using it. Taking a little extra time is definitely worth it.

**Comment by johanbrandhorst** [comment]
At the risk of becoming a tyrant on my first day, I'm going to commandeer and this and merge as @ivucica has been given ample time to add any more opinions and he conceded that he was happy with the PR even without the minor changes, that have been addressed.

**Comment by ivucica** [comment]
Sorry, travelling. No objection to the PR.

**Comment by achew22** [comment]
@johanbrandhorst I approve of the action and the attitude. Thanks for merging
